### DISCUSSION TITLE
JobManagementプロジェクトに含まれるライブラリ部品以外の分離検討

### DISCUSSION THREAD

**Discussion by makitagawa243** [Beckhoff-JP/PLC_JobManagementFramework#6] [General]
現状のJobManagementプロジェクトには、ライブラリ部品以外にも

* POUs/MAIN
* POUs/model/FB_ControllerType_1
* POUs/model/FB_ControllerType_1
* VISUs/Visualization

のPOU/VIsuが含まれています。
他プロジェクトで本Libraryを参照した際に、再利用することは無いであろう上記部品も一緒に付いてくるため
プロジェクトを分割して

* 純粋なライブラリ部品のみのプロジェクト
* ライブラリのデバッグ・実装例示用プロジェクト

に分けてはどうか、と考えました。
動機は以下の2点です。

1. FB_ControllerType_1/2でVAR_IN_OUT変数をメソッド内でアクセスしているため
https://infosys.beckhoff.com/english.php?content=../content/1033/tc3_plc_intro/2530307467.html&id=
に記載の警告が毎回出るのを回避したい（ライブラリから分離すれば警告は出なくなる）
1.  邪魔になるほどのボリュームでも無いが、純粋なライブラリにしておく方がユーザーは混乱しないのではないか

[テクニカルノート](https://beckhoff-jp.github.io/TwinCATHowTo/library/make_library.html#section-library-category)
を確認すると、

> とはいえ、動作実績のあるプログラムをそのままライブラリ化する事にも価値があります。なにより既存のPLCプロジェクトをそのままライブラリ化すれば良いだけなので手間がほどんど要りません。また、[ライブラリドキュメントの作成](https://beckhoff-jp.github.io/TwinCATHowTo/library/documentation.html#chapter-documentation)に示すreStructuredTextによるライブラリドキュメント機能を活用することでプログラム実装例の参考にも大いに役立ち、早期に開発者間での技術情報共有が可能になります。

上記での記載通り動くプログラム丸ごとをあえてライブラリに含めて公開する、というねらいでしょうか？
であれば、FB_ControllerType_1/2の警告回避とreStructuredTextでのコメント強化を進めればよいと考えます。

**Comment by ta-ichihashi** [comment]
投稿ありがとうございます。なるほど、気にしていませんでしたが、VAR_IN_OUTのせいで警告が出てしまうのですね。個人的にはある程度使い方を分かっている状態であれば、マニュアルよりも実装例を見た方が分かりやすくて再現が容易です。これに甘えてしまい、reStructuredTextをきっちり書くことをさぼりがちですね。
しかし、プロジェクトを分けてしまうと開発中のライブラリをテストする場合、毎回libraryとしてインストールしなければなりません。このわずらわしさは許容できないですね。本来は開発用のローカルプロジェクトをライブラリとして読み込みでき、編集が即座に反映される機能（Pythonの例だと"pip install -e"）が有ればよいのですが、今のところTwinCATでは見当たりません。（もしご存じならおしえてください）なので、仕方なく、ライブラリプロジェクトに実装サンプルを含めています。
テストフレームワークを導入する場合も、CI/CDの都合上、テストコードを同一プロジェクト内に同梱するのが通例なので、今のままで良いのではないでしょうか？
[](https://tcunit.org/#/programming-example-test-cases-part-one)

**Comment by makitagawa243** [comment]
> 毎回libraryとしてインストールしなければなりません。このわずらわしさは許容できない

これは同意見です。
一度試しにプロジェクトを分けてみましたが、
コード修正→ライブラリとして保存してインストール
を毎回実行する必要がありますし、デバッグ時にモニタするとライブラリ側のソースが開いてしまうので開発体験は非常に悪かったです。

.NETのライブラリでは、ライブラリ本体とテストプロジェクトがソリューション内に共存していればソリューション内のプロジェクト参照で解決するのですが、
TwinCATでは参照はどうやってもインストール済Libraryからを選ばざるを得ず、即時反映させる機能は見つけられていません。

思想は理解しましたので、警告回避のpull-requestを追って出します。
